pkg/sql/opt/memo/multiplicity_builder.go

```diff
@@ -331,16 +344,75 @@ func verifyFiltersAreValidEqualities(left, right RelExpr, filters FiltersExpr) b
 			rightTab = md.ColumnMeta(rightColID).Table
 			if leftTab == 0 || rightTab == 0 {
 				// Condition #4: Columns don't come from base tables.
-				return false
+				return opt.ColSet{}, false
 			}
 		}
 		if leftTab != md.ColumnMeta(leftColID).Table || rightTab != md.ColumnMeta(rightColID).Table {
 			// Condition #5: The filter conditions reference more than one table from
 			// each side.
-			return false
+			return opt.ColSet{}, false
 		}
+
+		rightEqualityCols.Add(rightColID)
 	}
-	return true
+
+	return rightEqualityCols, true
+}
+
+// leftAndRightColHaveMatchingFilters returns true if left and right are both
+// Select expressions that have an identical equality filters on leftCol and
+// rightCol, respectively.
+// TODO(mgartner): Extend this to find identical inequality filters. For
+// example, leftCol > 1 AND leftCol < 10, rightCol > 1 AND rightCol < 10.
+func leftAndRightColHaveMatchingFilters(left, right RelExpr, leftCol, rightCol opt.ColumnID) bool {
+	leftSelect, ok := left.(*SelectExpr)
+	if !ok {
+		return false
+	}
+	rightSelect, ok := right.(*SelectExpr)
+	if !ok {
+		return false
+	}
+
+	constValueForCol := func(filters FiltersExpr, col opt.ColumnID) (_ *ConstExpr, ok bool) {
+		var constant *ConstExpr
+		foundFilterWithCol := false
+		for i := range filters {
+			if !filters[i].ScalarProps().OuterCols.Contains(col) {
+				continue
+			}
+			if foundFilterWithCol {
+				// Ensure that this is the only filter on col. If there are
+				// others, we can't be sure that col is held to this constant.
+				// In practice, this should never occur because if there is an
+				// equality condition, the filters should be simplified so that
+				// there are no other filters on the same column.
+				return nil, false
+			}
+			foundFilterWithCol = true
+			eq, _ := filters[i].Condition.(*EqExpr)
+			if eq == nil {
+				continue
+			}
+			v, _ := eq.Left.(*VariableExpr)
+			c, _ := eq.Right.(*ConstExpr)
+			if v == nil || v.Col != col || c == nil {
+				continue
+			}
+			constant = c
+		}
+		return constant, constant != nil
+	}
+
+	leftConst, ok := constValueForCol(leftSelect.Filters, leftCol)
+	if !ok {
+		return false
+	}
+	rightConst, ok := constValueForCol(rightSelect.Filters, rightCol)
+	if !ok {
+		return false
+	}
+	return leftConst == rightConst
```

Comment:
I think I'm missing a check that the leftCol and rightCol are unfiltered in the inputs of the Select expressions.